Tests/kaas/kaas-sonobuoy-tests/scs_k8s_conformance_tests/scs_0217_cluster_hardening_test.go

```diff
@@ -0,0 +1,825 @@
+						if strings.Contains(cmd, "--authentication-token-webhook-config-file=") || strings.Contains(cmd, "--oidc-issuer-url=") {
+							authMethodsFound["OtherAuthMethod"] = true
+						}
+					}
+
+					// Check if both authentication methods are present
+					if authMethodsFound["ServiceAccountTokens"] && authMethodsFound["OtherAuthMethod"] {
+						t.Logf("Both authentication methods (Service Account Tokens and another method) are enabled")
+					} else {
+						if !authMethodsFound["ServiceAccountTokens"] {
+							t.Errorf("Error: Service Account Tokens are not enabled")
+						}
+						if !authMethodsFound["OtherAuthMethod"] {
+							t.Errorf("Error: No other authentication method (Token Webhook or OIDC) is enabled")
+						}
+					}
+				}
+			}
+
+			return ctx
+		})
+
+	testenv.Test(t, f.Feature())
+}
+
+// Test_scs_0217_etcd_tls checks if communication with etcd is secured with TLS for both peer- and cluster-communication.
+func Test_scs_0217_etcd_tls_communication(t *testing.T) {
+	f := features.New("etcd security").Assess(
+		"Communication with etcd MUST be secured with TLS for both peer- and cluster-communication",
+		func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			restConf, err := rest.InClusterConfig()
+			if err != nil {
+				t.Fatal("failed to create rest config:", err)
+			}
+
+			kubeClient, err := kubernetes.NewForConfig(restConf)
+			if err != nil {
+				t.Fatal("failed to create Kubernetes client:", err)
+			}
+
+			// Check kube-apiserver communication with etcd
+			checkKubeAPIServerETCDTLS(t, kubeClient)
+
+			return ctx
+		})
+
+	testenv.Test(t, f.Feature())
+}
+
+// Test_scs_0217_etcd_isolation checks ETCD is isolated from k8s cluster by checking the etcd server endpoints.
+func Test_scs_0217_etcd_isolation(t *testing.T) {
+	f := features.New("etcd security").Assess(
+		"ETCD should be isolated from the Kubernetes cluster",
+		func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			restConf, err := rest.InClusterConfig()
+			if err != nil {
+				t.Fatal("failed to create rest config:", err)
+			}
+
+			kubeClient, err := kubernetes.NewForConfig(restConf)
+			if err != nil {
+				t.Fatal("failed to create Kubernetes client:", err)
+			}
+
+			// Check if etcd is isolated from k8s cluster
+			checkIsolationETCD(t, kubeClient)
+
+			return ctx
+		})
+
+	testenv.Test(t, f.Feature())
+}
+
+// checkPortOpen tries to establish a TCP connection to the given IP and port.
+// It returns true if the port is open and false if the connection is refused or times out.
+func checkPortOpen(ip, port string, timeout time.Duration) bool {
+	address := net.JoinHostPort(ip, port)
+	conn, err := net.DialTimeout("tcp", address, timeout)
+	if err != nil {
+		// Connection failed, the port is likely closed
+		return false
+	}
+	// Connection succeeded, the port is open
+	conn.Close()
+	return true
+}
```

Comment:
There are some helper functions, please group them and put them in the beginning (please group the tests accordingly). That way, the tests are a bit more clearly structured. For reference see: https://github.com/SovereignCloudStack/standards/pull/714/files

Reply:
see 5ed2bfa9b43e1ae4d598d8576c7f5d79096c5fb2